### ISSUE TITLE
[Proposal] Shebang (#!) Support

### ISSUE THREAD

**Issue by 333fred** [dotnet/csharplang#3507]
# Shebang Directive

* [x] Proposed
* [ ] Prototype: [Complete](https://github.com/PROTOTYPE_OWNER/roslyn/BRANCH_NAME)
* [ ] Implementation: [In Progress](https://github.com/dotnet/roslyn/BRANCH_NAME)
* [ ] Specification: [Not Started](pr/1)

## Summary
[summary]: #summary

As we move to allow C# statements in the top level of a file, it's reasonable to assume that we might someday get support for `dotnet run`ing a .cs file. For further use of C# as a scripting language in this vein, I propose that we introduce a new preprocessor directive who's purpose is to exist solely to be ignored by the language: the shebang directive. Linux and Unix scripts will often use `#!<path to interpreter>` in order to direct the program loader to the interpreter that can run them. We should add a similar ignore to the C# language.

## Motivation
[motivation]: #motivation

Why are we doing this? What use cases does it support? What is the expected outcome?

## Detailed design
[design]: #detailed-design

We introduce a new directive:

```antlr
pp_directive
    : ... // Existing directives
    | pp_shebang
    ;

pp_shebang
    : '#' '!' input_character*
    ;
```

The language ignores the contents of the shebang directive and moves to the next line.

## Drawbacks
[drawbacks]: #drawbacks

Any change adds complication to the language. If we implement support for this and it starts appearing in files without actual scripting support, then it could be confusing for users expecting it to work.

## Alternatives
[alternatives]: #alternatives

N/A

## Unresolved questions
[unresolved]: #unresolved-questions

N/A

## Design meetings

N/A

**Comment by ufcpp** [comment]
I'm worried about BOM. [The presence of the BOM before the shebang will prevent the script interpreter](https://en.wikipedia.org/wiki/Shebang_(Unix)).

Visual Studio for Win uses Encoding.Default (usually not UTF8) if the BOM doesn't present.

**Comment by 333fred** [comment]
Sure, and we could certainly consider warning if the file uses anything other than UTF-8 w/o BOM.

**Comment by yaakov-h** [comment]
I thought shebang was already supported... is it only in the .csx variant?

**Comment by yaakov-h** [comment]
ah, looks like its only currently supported in the scripting mode.

https://github.com/dotnet/roslyn/blob/1b266a1e67434578ec09737a4e9427e52f44a420/src/Compilers/CSharp/Portable/Parser/DirectiveParser.cs#L108-L111

**Comment by svick** [comment]
1. As @yaakov-h mentions, shebang is already supported in the scripting dialect. Is this effectively formalizing that support and making it work in both dialects?
2. If I understand it correctly, the shebang that's supported by the scripting mode has to be literally the first thing in the file: no whitespace before it, no other lines before it. Since that's the only way shebang works, does it make sense to support whitespace before it and to support it on lines other than the first line in the file?

**Comment by HaloFour** [comment]
I'm a little confused.  Do "simple programs" not require a project file?  How would a single file "script" in this manner establish references or change any compiler properties?

This feels like trying to bridge a gap between C# "simple programs" and CSX, which already has this support plus the directive support to make it actually useful.  Short of adding that directive support this feels extremely limited.

Again, C# would benefit so much more from embracing CSX and improving the tooling support than it ever will be chasing CSX through some separate dialect that will always be much less capable.

**Comment by dersia** [comment]
I agree with @HaloFour without directives for assembly imports this might not be very useful. On the other hand having a script that downloads nuget packages or loading complex and somewhat big assemblies doesn't feel right for scripting.

The only thing that I kind of see as a big plus is that you can just take the cs file to any machine architecture and call `dotnet run` on it so you don't have to compile aot.

What is still unclear for me is if the purpose of this proposal is that the cs-file is being interpreted or is it going to be compiled on running it. 

If the purpose is to "just" interpreted on run, than I agree even more with @HaloFour we should focus on making csx a better experience.

**Comment by hez2010** [comment]
How about:
/// <interpreter path="" />

I don't think it should introduce a new `#!` directive to C#, for C# doesn't use `#` for comment syntax.

**Comment by HaloFour** [comment]
The point is to use a syntax that existing shells already recognize,
otherwise there would be no benefit.  Unix shells recognize shebang syntax
and it is widely used to enable a script file to instruct the shell in how
that file is to be interpreted.  No shell will understand a C# comment or a
blob of XML.

**Comment by GlassGrass** [comment]
I summarize my understanding of this proposal at followings. Is it right?

- C# compiler will simply ignore a shebang in the first line of the source. Because the lines start with `#` are preprocessor directives even on conventional spec, this modification looks not to impact other features.

- You can write inline csproj XML on and after 2nd line, but the interpretation of them should be done by build tool (e.g. dotnet-cli), not compiler. Such an inline csproj will be written under the conventional comment syntax, and the extension spec will be defined by tools.

```csharp
#!/bin/dotnet run
//+ <Project Sdk="Microsoft.NET.Sdk">
//+     <PropertyGroup>
//+         <TargetFramework>netcoreapp3.1</TargetFramework>
//+     </PropertyGroup>
//+ </Project>

/*
 * In this example, it is assumed that dotnet-cli will interprets the comments start with `//+` as inline csproj
 */

using System;

Console.WriteLine("Hello, world!");
```

**Comment by 333fred** [comment]
The first point is correct, and is the only thing this proposal covers. This proposal makes no statement on how a program would communicate dependencies to a tool.

**Comment by rolfbjarne** [comment]
One of the design meetings say:

> Would be nice to have, but only in conjunction with the rest of the simple programs feature. Implementation might be one of the simplest ever features in the compiler, but without the tooling to accompany it we don't know that this is the right design.

I'm not sure I agree, because this doesn't actually need official tooling to be useful, it's entirely possible to extend by anyone.

Consider this:

```cs
#!/bin/my-custom-program

using System;

Console.WriteLine("Hello, world!");
```

where `/bin/my-custom-program` can be a program that takes the script and does whatever it wants to - which could be creating a project file for the file and running `dotnet run`, but it an also be anything else.

One example could be:


```cs
#!/bin/run-unit-test

[TestFixture]
public class Test {
    [Test]
    public void IsValid (int number)
    {
        Assert.AreEqual (42, number, "everything");
    }
}
```

where `run-unit-test` would create a project, add the corresponding NUnit references and usings, compile & run the tests.

**Comment by CyrusNajmabadi** [comment]
> it's entirely possible to extend by anyone.

The point is that without that tooling we don't know if this will be good enough to actually solve those cases.

**Comment by rolfbjarne** [comment]
> > it's entirely possible to extend by anyone.
> 
> The point is that without that tooling we don't know if this will be good enough to actually solve those cases.

I agree that the reference directive (`#r`) might be in this category, but the shebang directive (`#!`) can stand on its own and would be useful by itself, both without the reference directive, or any other tooling.

**Comment by CyrusNajmabadi** [comment]
> > > it's entirely possible to extend by anyone.
> > 
> > The point is that without that tooling we don't know if this will be good enough to actually solve those cases.
> 
> I agree that the reference directive (`#r`) might be in this category, but the shebang directive (`#!`) can stand on its own and would be useful by itself, both without the reference directive, or any other tooling.

What is it's not good enough?  What if we do it, and a month later people try to use it and say it's not sufficient and it needs to change or be redone. 

We can't add features without some way of actually validating that it solves the problems we want it to solve. We need some actual tooling adoption to demonstrate it is sufficient.
